src/common/wal/src/lib.rs

```diff
@@ -30,3 +38,95 @@ pub enum TopicSelectorType {
     #[default]
     RoundRobin,
 }
+
+#[mockall::automock]
+#[async_trait]
+trait DNSResolver {
+    async fn resolve_broker_endpoint(&self, broker_endpoint: &str) -> io::Result<Vec<SocketAddr>>;
+}
+struct TokioDnsResolver;
+#[async_trait]
+impl DNSResolver for TokioDnsResolver {
+    async fn resolve_broker_endpoint(&self, broker_endpoint: &str) -> io::Result<Vec<SocketAddr>> {
+        net::lookup_host(broker_endpoint)
+            .await
+            .map(|iter| iter.collect())
+    }
+}
+async fn resolve_broker_endpoint_inner<R: DNSResolver>(
+    broker_endpoint: &str,
+    resolver: R,
+) -> Result<String> {
+    resolver
+        .resolve_broker_endpoint(broker_endpoint)
+        .await
+        .with_context(|_| ResolveEndpointSnafu {
+            broker_endpoint: broker_endpoint.to_string(),
+        })?
+        .into_iter()
+        // only IPv4 addresses are valid
+        .find(|addr| addr.is_ipv4())
+        .map(|addr| addr.to_string())
+        .with_context(|| EndpointIPV4NotFoundSnafu {
+            broker_endpoint: broker_endpoint.to_string(),
+        })
+}
+pub async fn resolve_broker_endpoint(broker_endpoint: &str) -> Result<String> {
+    resolve_broker_endpoint_inner(broker_endpoint, TokioDnsResolver).await
+}
```

Comment:
```suggestion
pub async fn resolve_to_ipv4(host: &str) -> Result<String> {
    tokio::net::lookup_host(host)
        .await
        .context(LookupHostSnafu { host })?
        .find_map(|s| s.is_ipv4().then_some(s.to_string()))
        .context(IPv4AddrNotFoundSnafu { host })
}
```

Reply:
Please ensure the `Cargo.toml` file will be cleaned accordingly since the `mockall` crate is removed.

Reply:
Your code is too compact that there're no new lines between functions. Please update your coding style to follow greptimedb.

Reply:
`with_context` is used when there's a potential heavy work to be done. Since snafu supports converting from &str to String, an explicit clone could be eliminated. Therefore, we can use `context` instead of `with_context`.

Reply:
> Your code is too compact that there're no new lines between functions. Please update your coding style to follow greptimedb.

I format the code with the `rustfmt.toml` in the root directory. I don't find any other style guide except the Pingcap style guide.